File: leetCode/matrix/flood-fill.go
// https://leetcode.com/problems/flood-fill/
type Pos struct {
    R int
    C int
}
func floodFillOldColor(image [][]int, sr int, sc int, newColor int, oldColor int) {
    //fmt.Printf("sr: %v, sc: %v, newColor: %v, oldColor: %v")
    if len(image) == 0 {
        return
    }
    lenR := len(image)
    lenC := len(image[0])
    var toCall []Pos
    if sr > 0 {
        /*if sc > 0 && image[sr - 1][sc - 1] == oldColor {
            toCall = append(toCall, Pos {R: sr - 1, C: sc - 1})
        }*/
        if image[sr - 1][sc] == oldColor {
            toCall = append(toCall, Pos {R: sr - 1, C: sc})
        }
        /*if sc + 1 < lenC && image[sr - 1][sc + 1] == oldColor {
            toCall = append(toCall, Pos {R: sr - 1, C: sc + 1})
        }*/
    }
    if sc > 0 && image[sr][sc - 1] == oldColor {
        toCall = append(toCall, Pos {R: sr, C: sc - 1})
    }
    if sc + 1 < lenC && image[sr][sc + 1] == oldColor {
        toCall = append(toCall, Pos {R: sr, C: sc + 1})
    }
    if sr + 1 < lenR {
        /*if sc > 0 && image[sr + 1][sc - 1] == oldColor {
            toCall = append(toCall, Pos {R: sr + 1, C: sc - 1})
        }*/
        if image[sr + 1][sc] == oldColor {
            toCall = append(toCall, Pos {R: sr + 1, C: sc})
        }
        /*if sc + 1 < lenC && image[sr + 1][sc + 1] == oldColor {
            toCall = append(toCall, Pos {R: sr + 1, C: sc + 1})
        }*/
    }
    image[sr][sc] = -newColor
    for _, pos := range toCall {
        floodFillOldColor(image, pos.R, pos.C, newColor, oldColor)
    }
}

func floodFill(image [][]int, sr int, sc int, newColor int) [][]int {
    floodFillOldColor(image, sr, sc, newColor, image[sr][sc])
    lenR := len(image)
    if lenR == 0 {
        return image
    }
    lenC := len(image[0])
    for i := 0; i < lenR; i++ {
        for j := 0; j < lenC; j++ {
            if image[i][j] < 0 {
                image[i][j] *= -1
            }
        }
    }
    return image
}

File: leetCode/matrix/flood-fill.rs
// https://leetcode.com/problems/flood-fill
use std::{
    cell::RefCell,
    rc::Rc,
};
impl Solution {
    fn fill(image: Rc<RefCell<Vec<Vec<i32>>>>, sr: i32, sc: i32, cur_color: i32) {
        if cur_color != image.borrow()[sr as usize][sc as usize] {
            return;
        }
        image.borrow_mut()[sr as usize][sc as usize] = -1;
        if sr - 1 >= 0 {
            Solution::fill(image.clone(), sr - 1, sc, cur_color);
        }
        if sc - 1 >= 0 {
            Solution::fill(image.clone(), sr, sc - 1, cur_color);
        }
        if (sr as usize + 1) < image.borrow().len() {
            Solution::fill(image.clone(), sr + 1, sc, cur_color);
        }
        if (sc as usize + 1) < image.borrow()[0].len() {
            Solution::fill(image.clone(), sr, sc + 1, cur_color);
        }
    }
    
    pub fn flood_fill(image: Vec<Vec<i32>>, sr: i32, sc: i32, color: i32) -> Vec<Vec<i32>> {
        let cur_color = image[sr as usize][sc as usize];
        let image = Rc::new(RefCell::new(image));
        Solution::fill(image.clone(), sr, sc, cur_color);

        let mut image = image.take();
        for row in &mut image {
            for x in row {
                if *x == -1 {
                    *x = color;
                }
            }
        }

        image
    }
}
